File: 2. adding two number - linked list/Solution.java
class Solution{
    class Node {
        int data;
        Node next;
        
        Node(int data) {
            this.data = data;
            this.next = null;
        }
    }
    class LinkedList {
        Node head;
        
        void insert(int data) {
            Node newNode = new Node(data);
            if (head == null) {
                head = newNode;
            } else {
                Node current = head;
                while (current.next != null) {
                    current = current.next;
                }
                current.next = newNode;
            }
        }
        
        void printList() {
            Node current = head;
            while (current != null) {
                System.out.print(current.data + " ");
                current = current.next;
            }
            System.out.println();
        }
    }
    public LinkedList addTwoNumbers(LinkedList l1, LinkedList l2) {
        LinkedList result = new LinkedList();
        Node p1 = l1.head;
        Node p2 = l2.head;
        int carry = 0;

        while (p1 != null || p2 != null || carry > 0) {
            int sum = carry;
            if (p1 != null) {
                sum += p1.data;
                p1 = p1.next;
            }
            if (p2 != null) {
                sum += p2.data;
                p2 = p2.next;
            }
            carry = sum / 10;
            result.insert(sum % 10);
        }

        return result;
    }
}

File: 2. adding two number - linked list/solution.js
class Node {
  constructor(value) {
    this.value = value;
    this.next = null;
  }
}
class LinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
  }

  append(value) {
    const newNode = new Node(value);
    if (!this.head) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.tail.next = newNode;
      this.tail = newNode;
    }
  }

  toArray() {
    const result = [];
    let current = this.head;
    while (current) {
      result.push(current.value);
      current = current.next;
    }
    return result;
  }
}

var addTwoNumbers = function(l1, l2) {
  const resultList = new LinkedList();
  let p = l1.head, q = l2.head;
  let carry = 0;

  while (p || q || carry) {
    const x = p ? p.value : 0;
    const y = q ? q.value : 0;
    const sum = x + y + carry;
    carry = Math.floor(sum / 10);
    resultList.append(sum % 10);
    if (p) p = p.next;
    if (q) q = q.next;
  }

  return resultList.toArray();
};